src/render.rs

```diff
@@ -310,18 +311,106 @@ impl Render for ServerNoticeMessageEventContent {
     }
 }
 
+/// Convert a matrix content URI to HTTP(s), respecting a user's homeserver
+fn mxc_to_http(
+    mxc_url: &str,
+    homeserver: &Url,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let url = url::Url::parse(mxc_url)?;
+
+    if url.scheme() != "mxc" {
+        return Err("URL missing MXC scheme".into());
+    }
+
+    if url.path().is_empty() {
+        return Err("URL missing path".into());
+    }
+
+    Ok(format!(
+        "{homeserver}_matrix/media/r0/download/{server_name}{media_id}",
+        homeserver = homeserver,
+        server_name = url.host().ok_or("Missing host")?,
+        media_id = url.path()
+    ))
+}
+
+/// Convert a matrix content URI to an encrypted mxc URI, respecting a user's homeserver.
+///
+/// The return value of this function will have a URI schema of emxc://. The path of the URI will
+/// be converted just like the mxc_to_http() function does, but it will also contain query
+/// parameters that are necessary to decrypt the payload the URI is pointing to.
+///
+/// This function is useful to present a clickable URI that can be passed to a plumber program that
+/// will download and decrypt the content that the matrix content URI is pointing to.
+///
+/// The returned URI should never be converted to http and opened directly, as that would expose
+/// the decryption parameters to any middleman or ISP.
+fn mxc_to_emxc(
+    mxc_url: &str,
+    homeserver: &Url,
+    encrypted: &EncryptedFile,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let url = url::Url::parse(mxc_url)?;
+
+    if url.scheme() != "mxc" {
+        return Err("URL missing MXC scheme".into());
+    }
+
+    if url.path().is_empty() {
+        return Err("URL missing path".into());
+    }
+
+    let host_str = format!(
+        "emxc://{}",
+        homeserver
+            .host_str()
+            .ok_or("Missing homeserver host string")?
+    );
+    let emxc_homeserver = url::Url::parse(&host_str)?;
+    let server_name = url.host().ok_or("Missing host")?;
+
+    let base_url = format!(
```

Comment:
Same as above, using join makes this a bit easier to follow and adds a bit more type safety.